api/src/main/java/org/pmiops/workbench/billing/FreeTierBillingService.java

```diff
@@ -227,132 +369,209 @@ private void maybeAlertOnCostThresholds(
 
-    return workspaceCosts;
+    return liveCostByWorkspace;
   }
 
   /**
-   * Retrieve the user's total free tier usage from the DB by summing across the Workspaces they
-   * have created. This is NOT live BigQuery data: it is only as recent as the last
-   * checkFreeTierBillingUsage cron job, recorded as last_update_time in the DB.
+   * Filter the costs further by getting the workspaces that are active, or deleted but their free
+   * tier last updated time is before the workspace last updated time. This filtration ensures that
+   * BQ will not be queried unnecessarily for the costs of deleted workspaces that we already have
+   * their latest costs The method will return the workspace in 1 of these cases:
    *
-   * <p>Note: return value may be null, to enable direct assignment to the nullable Profile field
+   * <p>1. The workspace is active 2. The workspace is deleted and either of the following is true
+   * 2.1. Free Tier Usage last updated time is null. This means that it wasn't calculated before
+   * 2.2. Workspace last updated time is null. This means we don't have enough info about the
+   * workspace, so we need to get its cost 2.3. Free Tier Usage time is before the Workspace last
+   * updated time. This means that the workspace got changed some time after our last calculation,
+   * so we need to recalculate it 2.4. Free Tier Usage time is after the Workspace last updated
+   * time, but the difference is smaller than a certain value. This case to account for charges that
+   * may occur after the workspace gets deleted and after the last cron had run.
    *
-   * @param user the user as represented in our database
-   * @return the total USD amount spent in workspaces created by this user, represented as a double
+   * @param allCostsInDbForUsers
+   * @return a {@link java.util.List} of the workspaces that require updates only
    */
-  @Nullable
-  public Double getCachedFreeTierUsage(DbUser user) {
-    return workspaceFreeTierUsageDao.totalCostByUser(user);
+  @NotNull
+  private List<WorkspaceCostView> findWorkspacesThatRequireUpdates(
+      List<WorkspaceCostView> allCostsInDbForUsers) {
+
+    List<WorkspaceCostView> workspacesThatRequireUpdate =
+        allCostsInDbForUsers.stream()
+            .filter(
+                c ->
+                    (c.getActiveStatus() == 0)
+                        || (c.getActiveStatus() == 1
+                            && (c.getFreeTierLastUpdated() == null
+                                || c.getWorkspaceLastUpdated() == null
+                                || c.getFreeTierLastUpdated().before(c.getWorkspaceLastUpdated())
+                                || (c.getFreeTierLastUpdated().after(c.getWorkspaceLastUpdated())
+                                    && c.getFreeTierLastUpdated().getTime()
+                                            - c.getWorkspaceLastUpdated().getTime()
+                                        < getNumberOfDaysToConsiderForFreeTierUsageUpdate()))))
```

Comment:
If we use Duration.ofDays() here. Then we may not need getNumberOfDaysToConsiderForFreeTierUsageUpdate method.